### PR TITLE
Implement views code for ES7.

### PR DESCRIPTION
**Pull request by dennisoelkers**
## Motivation and Context

This PR is adding the required code to support views for Elasticsearch
7. It consists of the views code from ES6 being copied over and adjusted
to use the related classes/methods from ES7.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/testing/TestMultisearchResponse.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class TestMultisearchResponse {
+    static List<NamedXContentRegistry.Entry> getDefaultNamedXContents() {
```
Comment (alex-konn) [inline graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/testing/TestMultisearchResponse.java:64]: All methods except `fromFixture` can be private. `fromFixture` should consequently be moved to the top of the class.

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        // on each nesting level and combination we have to check for series which we also add as values to the containing row
+        final HasAggregations initialResult = createInitialResult(queryResult);
+
+        processRows(resultBuilder, queryResult, queryContext, pivot, pivot.rowGroups(), new ArrayDeque<>(), initialResult);
+
+        return pivot.name().map(resultBuilder::name).orElse(resultBuilder).build();
+    }
+
+    private HasAggregations createInitialResult(SearchResponse queryResult) {
+        return InitialBucket.create(queryResult);
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java:232]: We do need `getDocCount()` in [one case](https://github.com/Graylog2/graylog2-server/blob/implement-views-for-es7/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/series/ESCountHandler.java#L72), so the best way seemed to be an implementation of `MultiBucketsAggregation.Bucket` which combines both `getDocCount()` and `HasAggregations`.

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+        // each combination of row and column groups can contain all series (if rollup is true)
+        // if rollup is false, only the "leaf" components contain the series
+        // in the elasticsearch result, rows and columns are simply nested aggregations (first aggregations from rows, then from columns)
+        // with metric aggregations on the corresponding levels.
+        // first we iterate over all row groups (whose values generate a "key array", corresponding to the nesting level)
+        // once we exhaust the row groups, we descend into the columns, which get added as values to their corresponding rows
+        // on each nesting level and combination we have to check for series which we also add as values to the containing row
+        final HasAggregations initialResult = createInitialResult(queryResult);
+
+        processRows(resultBuilder, queryResult, queryContext, pivot, pivot.rowGroups(), new ArrayDeque<>(), initialResult);
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java:226]: See https://github.com/Graylog2/graylog2-server/pull/8547#discussion_r459952287.

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/InitialBucket.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package org.graylog.storage.elasticsearch7.views.searchtypes.pivot;
+
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.action.search.SearchResponse;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.common.xcontent.XContentBuilder;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.search.aggregations.Aggregations;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.search.aggregations.bucket.MultiBucketsAggregation;
+
+import java.io.IOException;
+
+public class InitialBucket implements MultiBucketsAggregation.Bucket {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/InitialBucket.java:10]: See https://github.com/Graylog2/graylog2-server/pull/8547#discussion_r459952287.

[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/InitialBucket.java:10]: :+1:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java:232]: OK, got it. Sorry for missing that.

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/ESEventList.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+public class ESEventList implements ESSearchTypeHandler<EventList> {
+    @Override
+    public void doGenerateQueryPart(SearchJob job, Query query, EventList eventList,
+                                    ESGeneratedQueryContext queryContext) {
+        queryContext.searchSourceBuilder(eventList)
+                .size(10000);
+    }
+
+    @SuppressWarnings("unchecked")
+    protected List<Map<String, Object>> extractResult(SearchResponse result) {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/ESEventList.java:44]: `ESEventListTest` is overriding this method.

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendSearchTypeOverridesTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+        assertThat(indicesOf(generatedRequest))
+                .hasSize(2)
+                .containsExactly(
+                        "searchTypeIndex",
+                        "queryIndex"
+                );
+    }
+
+    private JsonPathAssert jsonAssertThat(DocumentContext ctx) {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendSearchTypeOverridesTest.java:170]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendGeneratedRequestTestBase.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    TimeRange timeRangeForTest() {
+        try {
+            return AbsoluteRange.create("2018-08-23T10:02:00.247+02:00", "2018-08-23T10:07:00.252+02:00");
+        } catch (InvalidRangeParametersException ignored) {
+        }
+        return null;
+    }
+
+    List<SearchRequest> run(SearchJob searchJob, Query query, ESGeneratedQueryContext queryContext, Set<QueryResult> predecessorResults) throws IOException {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendGeneratedRequestTestBase.java:128]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendGeneratedRequestTestBase.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        this.elasticsearchBackend.doRun(searchJob, query, queryContext, predecessorResults);
+
+        verify(client, times(1)).msearch(clientRequestCaptor.capture(), any());
+
+        final List<SearchRequest> generatedSearch = clientRequestCaptor.getValue();
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendGeneratedRequestTestBase.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendGeneratedRequestTestBase.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        return generatedSearch;
+    }
+
+    String resourceFile(String filename) {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendGeneratedRequestTestBase.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/series/ESCountHandler.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            record(queryContext, pivot, count, name, ValueCount.class);
+            return Optional.of(value);
+        }
+    }
+
+    @Override
+    public Stream<Value> doHandleResult(Pivot pivot,
+                                        Count count,
+                                        SearchResponse searchResult,
+                                        ValueCount metricAggregation,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/series/ESCountHandler.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/series/ESCountHandler.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        Object value = null;
+        if (metricAggregation instanceof MultiBucketsAggregation.Bucket) {
+            value = ((MultiBucketsAggregation.Bucket) metricAggregation).getDocCount();
+        } else if (metricAggregation instanceof Aggregations) {
+            value = searchResult.getHits().getTotalHits().value;
+        } else if (metricAggregation != null) {
+            value = metricAggregation.getValue();
+        }
+        if (value == null) {
+            LOG.error("Unexpected aggregation type {}, returning 0 for the count. This is a bug.", metricAggregation);
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/series/ESCountHandler.java]: I restructured it a bit and made the check for `valueCount`/`metricAggregation == null` more explicit.

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+
+        // holds the initial level aggregation to be added to the query
+        AggregationBuilder topLevelAggregation = null;
+        // holds the last complete bucket aggregation into which subsequent buckets get added
+        AggregationBuilder previousAggregation = null;
+
+        // add global rollup series if those were requested
+        if (pivot.rollup()) {
+            seriesStream(pivot, queryContext, "global rollup")
+                    .forEach(previousAggregation != null ? previousAggregation::subAggregation : searchSourceBuilder::aggregation);
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/ESEventList.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+public class ESEventList implements ESSearchTypeHandler<EventList> {
+    @Override
+    public void doGenerateQueryPart(SearchJob job, Query query, EventList eventList,
+                                    ESGeneratedQueryContext queryContext) {
+        queryContext.searchSourceBuilder(eventList)
+                .size(10000);
+    }
+
+    @SuppressWarnings("unchecked")
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/ESEventList.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ViewsESBackendModule.java [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    protected MapBinder<String, ESPivotSeriesSpecHandler<? extends SeriesSpec, ? extends Aggregation>> pivotSeriesHandlerBinder() {
+        return MapBinder.newMapBinder(binder(),
+                TypeLiteral.get(String.class),
+                new TypeLiteral<ESPivotSeriesSpecHandler<? extends SeriesSpec, ? extends Aggregation>>() {});
+
+    }
+
+    private ScopedBindingBuilder registerPivotSeriesHandler(
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ViewsESBackendModule.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ViewsESBackendModule.java [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    private MapBinder<String, ESPivotBucketSpecHandler<? extends BucketSpec, ? extends Aggregation>> pivotBucketHandlerBinder() {
+        return MapBinder.newMapBinder(binder(),
+                TypeLiteral.get(String.class),
+                new TypeLiteral<ESPivotBucketSpecHandler<? extends BucketSpec, ? extends Aggregation>>() {});
+
+    }
+
+    private ScopedBindingBuilder registerPivotBucketHandler(
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ViewsESBackendModule.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ViewsESBackendModule.java [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+
+    private LinkedBindingBuilder<RequestStrategy> bindRequestStrategy() {
+        return bind(RequestStrategy.class);
+    }
+
+    private LinkedBindingBuilder<ExportBackend> bindExportBackend() {
+        return bindExportBackend(SUPPORTED_ES_VERSION);
+    }
+
+    private void registerQueryBackend() {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ViewsESBackendModule.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendTestBase.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+package org.graylog.storage.elasticsearch7.views;
+
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.action.search.SearchRequest;
+import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+abstract class ElasticsearchBackendTestBase {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendTestBase.java:26]: :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/indexer/EventsIndexMappingBase.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.indexer;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+abstract class EventsIndexMappingBase {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog2/indexer/EventsIndexMappingBase.java:23]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/testing/TestMultisearchResponse.java (hunk 1 above)
Comment (dennisoelkers) [reply graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/testing/TestMultisearchResponse.java]: :heavy_check_mark:

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/ESEventList.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/ESEventList.java:43]: Oops. Right. Sorry I overlooked that.

**Pull request merged**